src/storage.rs

```diff
@@ -0,0 +1,215 @@
+mod schema;
+mod upgrade;
+
+use std::path::{Path, PathBuf};
+
+use crate::errors::*;
+
+pub struct Storage {
+    pub db_path: PathBuf,
+    pub root_path: PathBuf,
+    conn: rusqlite::Connection,
+}
+
+impl Storage {
+    pub fn create_at(db_path: &Path) -> Result<()> {
+        info!("Creating database at {}", db_path.display());
+        Self::create_or_open(db_path)?;
+        Ok(())
+    }
+
+    pub fn open(db_path: &Path) -> Result<Self> {
+        info!("Opening database at {}", db_path.display());
+        Self::create_or_open(db_path)
+    }
+
+    /// Open a sqlite3 DB file, also creating it if it doesn't already exist.
+    /// Note that the parent directory will NOT be created if it doesn't exist.
+    fn create_or_open(db_path: &Path) -> Result<Self> {
+        let conn = rusqlite::Connection::open(&db_path)
+            .map_err(|_| ErrorKind::DatabaseAccessError(db_path.to_path_buf()))?;
+
+        // Use a canonical path to avoid issues such as #168
+        let db_path = db_path
+            .canonicalize()
+            .map_err(|_| ErrorKind::NoDatabaseFound(db_path.to_path_buf()))?;
```

Comment:
The Go version used an absolute path. Using a canonical path instead should avoid #168.